snuba/datasets/errors_replacer.py

```diff
@@ -0,0 +1,427 @@
+import logging
+import time
+
+from collections import deque
+from datetime import datetime
+from typing import Any, Deque, FrozenSet, Mapping, Optional, Sequence, Tuple
+
+from snuba import settings
+from snuba.clickhouse import DATETIME_FORMAT
+from snuba.clickhouse.escaping import escape_identifier, escape_string
+from snuba.datasets.schemas.tables import TableSchema, WritableTableSchema
+from snuba.processor import InvalidMessageType, _hashify
+from snuba.redis import redis_client
+from snuba.replacers.replacer_processor import (
+    Replacement,
+    ReplacementMessage,
+    ReplacerProcessor,
+)
+
+
+logger = logging.getLogger(__name__)
+
+EXCLUDE_GROUPS = object()
+NEEDS_FINAL = object()
+
+
+def get_project_exclude_groups_key(project_id: int) -> str:
```

Comment:
I intentionally did not change much around the code organization of this class in order to make this PR as iso-functional as possible and make the review easier. So this is mostly cut/paste from replacer.py.
I just added better typing annotations.